File: mobile/test/household_mode_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:nurby_mobile/models/household_mode.dart';

void main() {
  group('ruleActiveIn', () {
    test('a rule with no modes fires in every mode', () {
      for (final m in kHouseholdModes) {
        expect(ruleActiveIn(null, m), isTrue);
        expect(ruleActiveIn({}, m), isTrue);
        expect(ruleActiveIn({'modes': <String>[]}, m), isTrue);
        expect(ruleActiveIn({'camera_id': 'x'}, m), isTrue);
      }
    });

    test('a gated rule fires only in the modes it names', () {
      final cond = {
        'modes': ['away', 'night']
      };
      expect(ruleActiveIn(cond, 'away'), isTrue);
      expect(ruleActiveIn(cond, 'night'), isTrue);
      expect(ruleActiveIn(cond, 'home'), isFalse);
    });

    test('a malformed modes value does not gate the rule', () {
      // Failing open here matters: a rule the client cannot parse must
      // keep firing, not go silently quiet.
      expect(ruleActiveIn({'modes': 'away'}, 'home'), isTrue);
    });
  });

  group('labels', () {
    test('an ungated rule has no gate label', () {
      expect(modeGateLabel(null), isNull);
      expect(modeGateLabel({'modes': <String>[]}), isNull);
      expect(modePausedLabel(null), isNull);
    });

    test('one mode', () {
      expect(modeGateLabel({'modes': ['away']}), 'Only while Away');
      expect(modePausedLabel({'modes': ['away']}), 'Paused until Away');
    });

    test('two modes read the way a person would say it', () {
      expect(modeGateLabel({'modes': ['away', 'night']}),
          'Only while Away or Night');
      expect(modePausedLabel({'modes': ['away', 'night']}),
          'Paused until Away or Night');
    });

    test('an unknown mode is dropped rather than shown raw', () {
      expect(modeGateLabel({'modes': ['away', 'vacation']}), 'Only while Away');
      expect(modeGateLabel({'modes': ['vacation']}), isNull);
    });

    test('every mode has a label', () {
      for (final m in kHouseholdModes) {
        expect(kModeLabels[m], isNotNull);
      }
    });
  });

  group('HouseholdModeState', () {
    final json = {
      'mode': 'away',
      'since': '2026-09-15T11:32:00.621051Z',
      'source': 'manual',
      'modes': [
        {'key': 'home', 'label': 'Home', 'hint': 'Someone is in.'},
        {'key': 'away', 'label': 'Away', 'hint': 'Nobody home.'},
        {'key': 'night', 'label': 'Night', 'hint': 'In for the night.'},
      ],
      'history': [
        {
          'id': 'abc',
          'mode': 'away',
          'previous_mode': 'home',
          'source': 'manual',
          'changed_by_name': 'Sam',
          'note': 'off to work',
          'changed_at': '2026-09-15T11:32:00.621051Z',
        }
      ],
      'silenced_rule_count': 2,
    };

    test('parses the server payload', () {
      final s = HouseholdModeState.fromJson(json);
      expect(s.mode, 'away');
      expect(s.silencedRuleCount, 2);
      expect(s.modes.map((m) => m.key), ['home', 'away', 'night']);
      expect(s.active?.label, 'Away');
      expect(s.history.single.changedByName, 'Sam');
      expect(s.history.single.note, 'off to work');
      expect(s.history.single.previousMode, 'home');
    });

    test('labels come from the server, not the local constants', () {
      // The point of serving them: a fourth mode added to the backend
      // renders without a client release.
      final s = HouseholdModeState.fromJson({
        ...json,
        'mode': 'holiday',
        'modes': [
          {'key': 'holiday', 'label': 'Holiday', 'hint': 'Away for a while.'}
        ],
      });
      expect(s.active?.label, 'Holiday');
      expect(s.active?.hint, 'Away for a while.');
    });

    test('survives a payload missing every optional field', () {
      final s = HouseholdModeState.fromJson({'mode': 'home'});
      expect(s.mode, 'home');
      expect(s.since, isNull);
      expect(s.modes, isEmpty);
      expect(s.history, isEmpty);
      expect(s.silencedRuleCount, 0);
    });

    test('defaults to home when the server sends nothing', () {
      expect(HouseholdModeState.fromJson({}).mode, 'home');
    });
  });
}

File: frontend/src/lib/__tests__/household-mode.test.ts
import { describe, it, expect } from "vitest";
import {
  HOUSEHOLD_MODES,
  MODE_LABELS,
  modeGateLabel,
  ruleActiveIn,
  type HouseholdMode,
} from "@/lib/household-mode";

describe("ruleActiveIn", () => {
  it("treats a rule with no modes as active everywhere", () => {
    for (const m of HOUSEHOLD_MODES) {
      expect(ruleActiveIn(null, m)).toBe(true);
      expect(ruleActiveIn(undefined, m)).toBe(true);
      expect(ruleActiveIn({}, m)).toBe(true);
      expect(ruleActiveIn({ modes: [] }, m)).toBe(true);
      expect(ruleActiveIn({ camera_id: "x" }, m)).toBe(true);
    }
  });

  it("gates a rule to only the modes it names", () => {
    const cond = { modes: ["away", "night"] };
    expect(ruleActiveIn(cond, "away")).toBe(true);
    expect(ruleActiveIn(cond, "night")).toBe(true);
    expect(ruleActiveIn(cond, "home")).toBe(false);
  });

  it("does not gate on a non-array modes value", () => {
    expect(ruleActiveIn({ modes: "away" }, "home")).toBe(true);
  });
});

describe("modeGateLabel", () => {
  it("returns null when the rule is not gated", () => {
    expect(modeGateLabel(null)).toBeNull();
    expect(modeGateLabel({})).toBeNull();
    expect(modeGateLabel({ modes: [] })).toBeNull();
  });

  it("names one mode", () => {
    expect(modeGateLabel({ modes: ["away"] })).toBe("Only while Away");
  });

  it("joins two modes the way a person would say it", () => {
    expect(modeGateLabel({ modes: ["away", "night"] })).toBe("Only while Away or Night");
  });

  it("drops a mode it does not recognise rather than printing the raw key", () => {
    expect(modeGateLabel({ modes: ["away", "vacation"] })).toBe("Only while Away");
    expect(modeGateLabel({ modes: ["vacation"] })).toBeNull();
  });
});

describe("the vocabulary", () => {
  it("labels every mode", () => {
    for (const m of HOUSEHOLD_MODES) {
      expect(MODE_LABELS[m as HouseholdMode]).toBeTruthy();
    }
  });
});

describe("modePausedLabel", () => {
  it("names the mode that would wake the rule, so the badge is actionable", async () => {
    const { modePausedLabel } = await import("@/lib/household-mode");
    expect(modePausedLabel({ modes: ["away"] })).toBe("Paused until Away");
    expect(modePausedLabel({ modes: ["away", "night"] })).toBe("Paused until Away or Night");
  });

  it("keeps the mode names capitalised", async () => {
    const { modePausedLabel } = await import("@/lib/household-mode");
    expect(modePausedLabel({ modes: ["away", "night"] })).toContain("Away");
    expect(modePausedLabel({ modes: ["away", "night"] })).toContain("Night");
  });

  it("returns null for an ungated rule", async () => {
    const { modePausedLabel } = await import("@/lib/household-mode");
    expect(modePausedLabel(null)).toBeNull();
    expect(modePausedLabel({ modes: [] })).toBeNull();
  });
});
